File: servers/Cour-Management/src/main/java/com/radim/project/entity/Lesson.java
package com.radim.project.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "lessons")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Lesson {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "module_id", nullable = false)
    @ToString.Exclude
    private Module module;

    @NotBlank
    private String title;

    private String summary;

    private int orderIndex;

    @OneToMany(mappedBy = "lesson", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<LessonContent> contents;

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: clients/mobile/lib/data/models/course/lesson.dart
import 'lesson_content.dart' show LessonContent;

class Lesson {
  final String id;
  final String moduleId;
  final String title;
  final String? summary;
  final int orderIndex;
  final List<LessonContent>? contents;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Lesson({
    required this.id,
    required this.moduleId,
    required this.title,
    this.summary,
    required this.orderIndex,
    this.contents,
    this.createdAt,
    this.updatedAt,
  });

  factory Lesson.fromJson(Map<String, dynamic> json) {
    return Lesson(
      id: json['id'] as String,
      moduleId: json['moduleId'] as String,
      title: json['title'] as String,
      summary: json['summary'] as String?,
      orderIndex: json['orderIndex'] as int,
      contents: json['contents'] != null
          ? (json['contents'] as List)
              .map((c) => LessonContent.fromJson(c))
              .toList()
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'moduleId': moduleId,
      'title': title,
      if (summary != null) 'summary': summary,
      'orderIndex': orderIndex,
      if (contents != null)
        'contents': contents!.map((c) => c.toJson()).toList(),
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
    };
  }
}

class CreateLessonRequest {
  final String title;
  final String? summary;
  final int orderIndex;

  CreateLessonRequest({
    required this.title,
    this.summary,
    required this.orderIndex,
  });

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      if (summary != null) 'summary': summary,
      'orderIndex': orderIndex,
    };
  }
}

